### ISSUE TITLE
MongoDB pod is a single point of failure in default deploy

### ISSUE THREAD

**Issue by absoludity** [kubeapps/kubeapps#1141]
From @migmartri:

We noticed an outage at 23:02:02Z with the following message from TIller proxy
```
time="2019-08-25T23:02:02Z" level=fatal msg="unable to connect to MongoDB" host=kubeapps-mongodb
```
When I check the logs for Mongo, it indeed seem to have been initialized around that time, which will explain not being to connect.

```
k logs kubeapps-mongodb-679dc7dcdf-bbzfr 

Welcome to the Bitnami mongodb container
Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-mongodb
Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-mongodb/issues

nami    INFO  Initializing mongodb
mongodb INFO  ==> Deploying MongoDB from scratch...
mongodb INFO  ==> No injected configuration files found. Creating default config files...
mongodb INFO  ==> Creating root user...
mongodb INFO  ==> Enabling authentication...
mongodb INFO 
mongodb INFO  ########################################################################
mongodb INFO   Installation parameters for mongodb:
mongodb INFO     Root Password: **********
mongodb INFO   (Passwords are not shown for security reasons)
mongodb INFO  ########################################################################
mongodb INFO 
nami    INFO  mongodb successfully initialized
INFO  ==> Starting mongodb... 
INFO  ==> Starting mongod...
2019-08-25T23:01:41.951+0000 I CONTROL  [main] Automatically disabling TLS 1.0, to force-enable TLS 1.0 specify --sslDisabledProtocols 'none'
2019-08-25T23:01:41.966+0000 I CONTROL  [initandlisten] MongoDB starting : pid=92 port=27017 dbpath=/opt/bitnami/mongodb/data/db 64-bit host=kubeapps-mongodb-679dc7dcdf-bbzfr
2019-08-25T23:01:41.966+0000 I CONTROL  [initandlisten] db version v4.0.3
2019-08-25T23:01:41.966+0000 I CONTROL  [initandlisten] git version: 7ea530946fa7880364d88c8d8b6026bbc9ffa48c
2019-08-25T23:01:41.966+0000 I CONTROL  [initandlisten] OpenSSL version: OpenSSL 1.1.0j  20 Nov 2018
```
But the weird thing is that the MongoDB pod has been running for more than 2 days and no restarts. How is that possible? What am I reading wrong?

```
k get pods -l release=kubeapps
kubeapps-mongodb-679dc7dcdf-bbzfr                            1/1     Running   0          2d15h
```

**Comment by absoludity** [comment]
After investigating, I don't believe this issue is related to the config of MongeDB within Kubeapps, nor Kubeapps itself, but rather is a cluster health issue.

I see logs such as:
```
~$ k -n kubeapps logs --previous kubeapps-internal-tiller-proxy-6db755cdc4-7w7gq
time="2019-08-28T00:12:40Z" level=info msg="Using tiller host: tiller-deploy.kube-system:44134"
time="2019-08-28T00:12:45Z" level=fatal msg="Unable to connect to Tiller: context deadline exceeded"
```

and more generally, the nodes of the cluster are cycling quite regularly - checking them now shows:
```
~$ k get nodes -o jsonpath='{range .items[*]}{.metadata.name}{"\t"}{.metadata.creationTimestamp}{"\n"}{end}'
gke-[cluster-name]-default-pool-0d14fd5a-2vn6   2019-08-26T13:54:14Z
gke-[cluster-name]-default-pool-0d14fd5a-7xv6   2019-08-27T08:13:21Z
gke-[cluster-name]-default-pool-0d14fd5a-tb1x   2019-08-26T01:28:02Z
gke-[cluster-name]-default-pool-ac32caac-dn4l   2019-08-26T21:17:02Z
gke-[cluster-name]-default-pool-ac32caac-j6hx   2019-08-26T23:12:40Z
gke-[cluster-name]-default-pool-ac32caac-t6b4   2019-08-27T22:13:28Z
gke-[cluster-name]-default-pool-e8e08f57-3k34   2019-08-27T09:47:28Z
gke-[cluster-name]-default-pool-e8e08f57-rztf   2019-08-28T00:12:38Z
gke-[cluster-name]-default-pool-e8e08f57-wwrl   2019-08-25T23:08:30Z
```

Checking for activity (after figuring out gcloud logging incantations) right before the time of the incident in the description: 2019-08-25T23:00:00.000Z, I see that the event correlates with a failed node creation API call, but the call fails because the node already exists (which seemed odd at first, a CREATE api call for a named resource that we know to exist):
```
~$ gcloud logging read --project $PROJECT_NAME "logName=\"projects/$PROJECT_NAME/logs/cloudaudit.googleapis.com%2Factivity\" AND resource.type=\"k8s_cluster\" AND resource.labels.cluster_name=\"$CLUSTER_NAME\" AND (protoPayload.methodName=\"io.k8s.core.v1.nodes.create\" OR protoPayload.methodName=\"io.k8s.core.v1.nodes.delete\") AND timestamp>=\"2019-08-25T22:55:00.000000Z\" AND timestamp<=\"2019-08-25T23:05:00.000000Z\"" --format json | jq '.[] | {timestamp: .timestamp, methodName: .protoPayload.methodName, status: .protoPayload.response.status, reason: .protoPayload.response.reason}'
{
  "timestamp": "2019-08-25T23:00:55.050350Z",
  "methodName": "io.k8s.core.v1.nodes.create",
  "status": "Failure",
  "reason": "AlreadyExists"
}
```

But it looks like this request is just the node controller testing if the node is available again, because checking patch requests, I see the node was tainted around 1 minute earlier and then taints removed milliseconds after the above create call:
```
~$ gcloud logging read --project $PROJECT_NAME "logName=\"projects/$PROJECT_NAME/logs/cloudaudit.googleapis.com%2Factivity\" AND resource.type=\"k8s_cluster\" AND resource.labels.cluster_name=\"$CLUSTER_NAME\" AND protoPayload.methodName=\"io.k8s.core.v1.nodes.patch\" AND timestamp>=\"2019-08-25T22:55:00.000000Z\" AND timestamp<=\"2019-08-25T23:05:00.000000Z\"" --format json | jq '.[] | {timestamp: .timestamp, methodName: .protoPayload.methodName, nodeName: .protoPayload.resourceName, spec: .protoPayload.request.spec}'
{
  "timestamp": "2019-08-25T23:00:55.625030Z",
  "methodName": "io.k8s.core.v1.nodes.patch",
  "nodeName": "core/v1/nodes/gke-andres-cluster-default-pool-0d14fd5a-nft6",
  "spec": {
    "taints": null
  }
}
{
  "timestamp": "2019-08-25T22:59:52.211236Z",
  "methodName": "io.k8s.core.v1.nodes.patch",
  "nodeName": "core/v1/nodes/gke-andres-cluster-default-pool-0d14fd5a-nft6",
  "spec": {
    "taints": [
      {
        "effect": "NoSchedule",
        "key": "node.kubernetes.io/unreachable",
        "timeAdded": "2019-08-25T22:59:52Z"
      }
    ]
  }
}
```

So, at best, we know a node was not responding in the same minute that the failure was observed for mongodb. I was unable to find more useful information such as whether the node was drained, causing the mongodb pod to be moved to a new node, which could *possibly* explain seeing the container starting on a 2 day old pod without restarts (though I haven't confirmed this).

Checking the other time that Miguel mentioned privately, 2019-08-27T22:16:00Z, shows that at 22:12 a node similarly went unavailable:
```
~$ gcloud logging read --project $PROJECT_NAME "logName=\"projects/$PROJECT_NAME/logs/cloudaudit.googleapis.com%2Factivity\" AND resource.type=\"k8s_cluster\" AND resource.labels.cluster_name=\"$CLUSTER_NAME\" AND protoPayload
.methodName=\"io.k8s.core.v1.nodes.patch\" AND timestamp>=\"2019-08-27T22:10:00.000000Z\" AND timestamp<=\"2019-08-27T22:20:00.000000Z\"" --format json | jq '.[] | {timestamp: .timestamp, methodName: .protoPayload.methodName, 
nodeName: .protoPayload.resourceName, spec: .protoPayload.request.spec}'                                                                                                                                                          
{
  "timestamp": "2019-08-27T22:13:47.030925Z",
  "methodName": "io.k8s.core.v1.nodes.patch",
  "nodeName": "core/v1/nodes/gke-andres-cluster-default-pool-ac32caac-t6b4",
  "spec": {
    "taints": null
  }
}
{
  "timestamp": "2019-08-27T22:13:28.958613Z",
  "methodName": "io.k8s.core.v1.nodes.patch",
  "nodeName": "core/v1/nodes/gke-andres-cluster-default-pool-ac32caac-t6b4",
  "spec": {
    "taints": [
      {
        "effect": "NoSchedule",
        "key": "node.kubernetes.io/network-unavailable",
        "timeAdded": "2019-08-27T22:13:28Z"
      }
    ]
  }
}
{
  "timestamp": "2019-08-27T22:13:28.879035Z",
  "methodName": "io.k8s.core.v1.nodes.patch",
  "nodeName": "core/v1/nodes/gke-andres-cluster-default-pool-ac32caac-t6b4",
  "spec": null
}
{
  "timestamp": "2019-08-27T22:13:28.862955Z",
  "methodName": "io.k8s.core.v1.nodes.patch",
  "nodeName": "core/v1/nodes/gke-andres-cluster-default-pool-ac32caac-t6b4",
  "spec": {
    "taints": [
      {
        "effect": "NoSchedule",
        "key": "node.kubernetes.io/network-unavailable",
        "timeAdded": "2019-08-27T22:13:28Z"
      },
      {
        "effect": "NoSchedule",
        "key": "node.kubernetes.io/not-ready",
        "timeAdded": "2019-08-27T22:13:28Z"
      }
    ]
  }
}
{
  "timestamp": "2019-08-27T22:13:28.845472Z",
  "methodName": "io.k8s.core.v1.nodes.patch",
  "nodeName": "core/v1/nodes/gke-andres-cluster-default-pool-ac32caac-t6b4",
  "spec": {
    "podCIDR": "10.28.35.0/24"
  }
}
{
  "timestamp": "2019-08-27T22:13:28.835777Z",
  "methodName": "io.k8s.core.v1.nodes.patch",
  "nodeName": "core/v1/nodes/gke-andres-cluster-default-pool-ac32caac-t6b4",
  "spec": null
}
{
  "timestamp": "2019-08-27T22:12:31.103260Z",
  "methodName": "io.k8s.core.v1.nodes.patch",
  "nodeName": "core/v1/nodes/gke-andres-cluster-default-pool-ac32caac-t6b4",
  "spec": {
    "taints": [
      {
        "effect": "NoSchedule",
        "key": "node.kubernetes.io/unreachable",
        "timeAdded": "2019-08-27T22:12:31Z"
      }
    ]
  }
}
```

Oddly, at this time it does appear that the API is used to *delete* the node, before requesting creating the same node:
```
~$ gcloud logging read --project $PROJECT_NAME "logName=\"projects/$PROJECT_NAME/logs/cloudaudit.googleapis.com%2Factivity\" AND resource.type=\"k8s_cluster\" AND resource.labels.cluster_name=\"$CLUSTER_NAME\" AND (protoPayload.methodName=\"io.k8s.core.v1.nodes.create\" OR protoPayload.methodName=\"io.k8s.core.v1.nodes.delete\") AND timestamp>=\"2019-08-27T22:10:00.000000Z\" AND timestamp<=\"2019-08-27T22:20:00.000000Z\"" --format json | jq '.[] | {timestamp: .timestamp, methodName: .protoPayload.methodName, resourceName: .protoPayload.resourceName, status: .protoPayload.status}'
{
  "timestamp": "2019-08-27T22:13:28.815364Z",
  "methodName": "io.k8s.core.v1.nodes.create",
  "resourceName": "core/v1/nodes/gke-andres-cluster-default-pool-ac32caac-t6b4",
  "status": {}
}
{
  "timestamp": "2019-08-27T22:12:58.286601Z",
  "methodName": "io.k8s.core.v1.nodes.delete",
  "resourceName": "core/v1/nodes/gke-andres-cluster-default-pool-ac32caac-t6b4",
  "status": {}
}
```

And this create timestamp for the API call matches the `-t6b4` creation timestamp that I happened to capture hours ago above, so it indeed appears that the nodes are being re-created quite often. We just don't yet know why.

**Comment by absoludity** [comment]
One thing I forgot to mention @andresmgot :
```
~$ gcloud container node-pools describe default-pool --cluster $CLUSTER_NAME --project $PROJECT_NAME --region us-east1 --format json | jq '.management'
{
  "autoRepair": true
}
```

So the re-creation of the nodes could be the autoRepair in action, but we would want to know why the nodes are becoming unreachable (or otherwise being replaced by autoRepair).

Also, perhaps we can close this task and start a doc or similar, since it appears unrelated to Kubeapps or mongodb?

**Comment by andresmgot** [comment]
I can confirm that the issue is caused by the auto-recreation of nodes. The recreation criteria is (from the [docs](https://cloud.google.com/kubernetes-engine/docs/how-to/node-auto-repair)) that a node  doesn't report a healthy status in less than 10 min or the boot disk is full (for <30min).

I am not sure what's the reason of the node becoming unhealthy because the actual deployments doesn't report any issue regarding the pods being unhealthy and since the nodes got deleted, I am not able to retrieve any log or status. The only information of the event I got is:

```
statusMessage: "Instance Group Manager 'projects/<>/zones/us-east1-d/instanceGroupManagers<>-cluster-default-pool-0d14fd5a-grp'\
  \ initiated recreateInstance on instance 'projects/<>/zones/us-east1-d/instances/<>-cluster-default-pool-0d14fd5a-2vn6'.\
  \ Reason: expected instance status RUNNING but STOPPING found."
```

I see that even the resource usage is not very high but the requested memory and cpu for some nodes is quite high:

```
| Name | Status | CPU requested | CPU allocatable | Memory requested | Memory allocatable | Storage requested | Storage allocatable |  
| gke-andres-cluster-default-pool-0d14fd5a-2vn6 | Ready | 926 mCPU | 940 mCPU | 2.06 GB | 2.77 GB
```

I have increased the number of nodes as immediate fix. This will increase the amount of resources available.

Since we have more resources I think we can turn off the _autorepair_ behavior. It seems weird to me that a Node is not healthy when none of the Pods within it report any issue. From the events you collect it seems that the node is unreachable for just 1 minute (not the 10 minutes that gcloud claims). Also, with this we can evaluate what's happening with the node in case it is really wrong. Since we have plenty of other nodes this should not be a problem, pods should be able to be rescheduled in a different node.

In any case, it's true that this issue is not directly related to Kubeapps but to this specific cluster environment. Note that not only the `mongodb` deployment got rescheduled but also `tiller`. That's what caused the downtime for Tiller proxy (since that service doesn't depend on MongoDB).

The only thing we can do on our side (Kubeapps) to avoid this kind of issues is to reduce the dependency with MongoDB (the dependency with Tiller will go away with Helm v3). We can do so storing a cache of the results of MongoDB in the `chartsvc` so we can still return results during transient downtime of MongoDB.

**Comment by absoludity** [comment]
> I can confirm that the issue is caused by the auto-recreation of nodes. The recreation criteria is (from the [docs](https://cloud.google.com/kubernetes-engine/docs/how-to/node-auto-repair)) that a node doesn't report a healthy status in less than 10 min or the boot disk is full (for <30min).
> 
> I am not sure what's the reason of the node becoming unhealthy because the actual deployments doesn't report any issue regarding the pods being unhealthy and since the nodes got deleted, I am not able to retrieve any log or status. The only information of the event I got is:
> 
> ```
> statusMessage: "Instance Group Manager 'projects/<>/zones/us-east1-d/instanceGroupManagers<>-cluster-default-pool-0d14fd5a-grp'\
>   \ initiated recreateInstance on instance 'projects/<>/zones/us-east1-d/instances/<>-cluster-default-pool-0d14fd5a-2vn6'.\
>   \ Reason: expected instance status RUNNING but STOPPING found."
> ```

So the google cloud instance is in a STOPPING state - that sounds like a good reason for k8s to be recreating the node worker?

> 
> I see that even the resource usage is not very high but the requested memory and cpu for some nodes is quite high:

Right, from the info you found above, it's unlikely to be related to the resources on the node, but something explicitly stopping the nodes. Perhaps a lack of required tags and it's being auto-reaped?

> I have increased the number of nodes as immediate fix. This will increase the amount of resources available.

I really don't think this will help. 9 worker nodes should already be ample. Given all have been restarted in the last 3 days, if that's because the cloud instance was "STOPPING" as you found above, I'd look for what's stopping the instances. We should be able to see that.

> Since we have more resources I think we can turn off the _autorepair_ behavior. It seems weird to me that a Node is not healthy when none of the Pods within it report any issue. From the events you collect it seems that the node is unreachable for just 1 minute (not the 10 minutes that gcloud claims).

Probably not the autorepair (ie. the node wasn't unhealthy - the cloud instance was stopped). So I think we'll find they are still recreated to maintain the number of nodes (as opposed to the auto repair functionality).

**Comment by andresmgot** [comment]
> Given all have been restarted in the last 3 days, if that's because the cloud instance was "STOPPING" as you found above, I'd look for what's stopping the instances. We should be able to see that.

Good point. I realized that nodes were running in `preemptible` nodes, which caused GKE to remove those machine after certain time. I have created a different pool of nodes and tainted the existing ones to avoid new pods to be scheduled in those. I will remove them once they are empty.

**Comment by absoludity** [comment]
> Note that during these days I've been increasing/decreasing the number of nodes.

Yep, unfortunately due to the above, this won't have helped (unless you also happened to move kubeapps pods onto the new nodes which were safe for the duration of your work). Though perhaps it increased the chances of pods for new deploys of charts being on nodes that weren't about to be terminated.

> 
> In any case, I think that Kubeapps could be more resilient to these kind of changes, at the end of the day, pod re-shuffling is an expected behavior in Kubernetes and can happen not only because of a node issue but because memory/compute constraints.

Definitely.

> 
> From the App POV I think that the outage looks like this:
> 
> 1 - MongoDB pod get's re-scheduled, initialization will take x minutes.

But isn't the issue here is that MongoDB is currently a single point of failure? (as Redis or anything else would be if it was deployed with only a single replica). Reading back it looks like this is because it's considered to be more of a cache than storage ([see comment in values related to persistence which says as much and defaults fileStorage to false](https://github.com/kubeapps/kubeapps/blob/master/chart/kubeapps/values.yaml#L257)). The frontend, dashboard, chartsvc and tiller-proxy all have a replica count of 2 by default mongo DB does not (defaulting to a single pod).

> 2 - Charts service will crash but eventually recover.

I assume this would continue to serve data if the backend data was not a single pod?

> 3 - No charts will be available since we rely on appRepository K8s cronjobs. That means that worst case scenario we will have no content for [refresh cycle] minutes.

Again, with a replica set, we would not lose data.

> 
> I'd like to explore:
> 
> a - Do we have persistence in MongoDB? If we do not why? That will save us from 3.
As above, persistence is explicitly defaulting to false with a comment along the lines of mongodb being a cache.

> b - Are we sure that the MongoDB pod did not get re-scheduled because of memory/CPU constraints in the cluster and the pod running way over it's `request` settings?

In your cases investigated above, the outage definitely correlates within seconds of a k8s node being recreated.

> Currently its QOS is burstable and we should double check that requests/limits are properly set. In other words, the `requested` memory/cpu is reasonable and the operating resources are around that ballpark.

Sounds worth checking either way.

In summary, though the reason persistence is defaulting to false may no longer be true (that is, even though it is a cache, it is by default a SPOF cache causing outages when rescheduled), I think the primary solution should be to ensure it is not a SPOF by default (replica count) as we would for anything data storage with which mongodb was replaced? It'll be easy to test this (ie. using at least two replicas and then forcing one to be rescheduled).

**Comment by absoludity** [comment]
> > I assume this would continue to serve data if the backend data was not a single pod?
> 
> I think that enabling persistence will help even in the case of having a single pod, sure we will have downtime, but much shorted since the new pod MongoDB initialization will be skipped and the data will be served right away. Today, our outage is not only problematic because of the initialization of the DB itself, but also because of the data being loaded asynchronously (via k8s cronjobs), in periods of 15 minutes (I think).

Oh it'll definitely help, but (a) in the same way that enabling replica sets will (ie. data is available if one pod is rescheduled), and (b) it will remain a SPOF. I'm just saying I'd prioritise replicasets, then persistence. But both would be better, sure.

> > Reading back it looks like this is because it's considered to be more of a cache than storage (see comment in values related to persistence which says as much and defaults fileStorage to false).
> 
> But I think this is a wrong assumption, since the ChartSVC requires an async job to populate "this cache". If we were making it a cache, the index of the charts and its retrieval should be sync (and then cached).

Yeah, I agree :)

Closing again and I'll later create a separate task or two for moving forward. Thanks Migue.
